Add possibility do data define label position by point additionally to X/Y

domi4484 opened the pull request:
This will allow to use the move label tool to change the position of a label coming from a point geometry column (It is assumed that the additional geometry has the same CRS that the main one).
It will also be possible do define just one expression for the label position and not two very similar expressions for X and Y.
Activating the Point override expression will deactivate X/Y overrides, and activating X or Y will deactivate Point.

Gui changes:
![image](https://user-images.githubusercontent.com/9881900/144635934-8622cf3b-4246-4b2d-b4a8-e373359d414c.png)

domi4484 committed:
switch to XY coordinate types when using label move tool

domi4484 committed:
remove unneeded SIP_CONVERT_TO_SUBCLASS_CODE

domi4484 committed:
Add possibility do data define label position by point instead of X/Y

domi4484 committed:
added missing QRegularExpression include

domi4484 committed:
added mask image

domi4484 committed:
Removed ewkt conversion functions

domi4484 committed:
Removed ewkt conversion test

domi4484 committed:
Simplified configuration gui

domi4484 committed:
reverted changes added by error on merge

nyalldawson commented:
I'd like to review before.merge

roya0045 commented:
Would this be where the values be set when moving labels?

How does it handle the copy/paste symbology? Sometimes the x & Y fields become invalidated and using free placement doesn't work and the values need to be reset.

m-kuhn commented on python/gui/auto_generated/qgstextformatwidget.sip.in:
```diff
@@ -145,11 +145,9 @@ Sets the background color for the text preview widget.
 %End
 
-    void enableDataDefinedAlignment( bool enable );
+    void updateDataDefinedAlignment();
```
@nyalldawson what is the current approach with API changes like this (not often used)?
Can they be documented in api_break.dox or do we need to keep a `( bool enable = false)` and document it as unused param to be removed with QGIS 4?

nyalldawson replied:
I'd suggest deprecating the old enableDataDefinedAlignment method and making it a no-op. 

(And then hiding the new updateDataDefinedAlignment() method from sip, as neither should be exposed)

m-kuhn commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
+          QgsGeometry geometryPoint = pointAsVariant.value<QgsGeometry>();
+          const QgsPoint *point  = qgsgeometry_cast<QgsPoint *>( geometryPoint.constGet() );
+
+          x = point->x();
```
Can you check the point for nou null here?

m-kuhn commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
+          xSuccess = true;
+          ySuccess = true;
+        }
+        else if ( !pointAsVariant.toByteArray().isEmpty() )
```
Where are bytearray wkb points produced?

nyalldawson replied:
I'm guessing it's to support (Read only) fields which are already created on the database as wkb.

Personally I don't think this a safe approach -- we're feeding any random binary field into the fromWkb method, without any safeguard that it is actually wkb content. 

I think we could drop this handling here and instead just get users to use an expression based property with a `geom_from_wkb( "wkb_field")` expression if they need this functionality.

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
+        if ( pointAsVariant.canConvert<QgsGeometry>() )
+        {
+          QgsGeometry geometryPoint = pointAsVariant.value<QgsGeometry>();
+          const QgsPoint *point  = qgsgeometry_cast<QgsPoint *>( geometryPoint.constGet() );
```
Hmm, this really shouldn't compile at all -- it should be complaining because the const pointer is being cast to a non-const pointer :scream:

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
+        QVariant pointAsVariant = attributes.at( pointCol );
+        if ( pointAsVariant.canConvert<QgsGeometry>() )
+        {
+          QgsGeometry geometryPoint = pointAsVariant.value<QgsGeometry>();
+          const QgsPoint *point  = qgsgeometry_cast<QgsPoint *>( geometryPoint.constGet() );
+
+          x = point->x();
+          y = point->y();
+
+          xSuccess = true;
+          ySuccess = true;
```
```suggestion
         const  QgsGeometry geometry = pointAsVariant.value<QgsGeometry>();
         if ( const QgsPoint *point  = ( geometry.constGet() ? qgsgeometry_cast<QgsPoint *>( geometry.constGet()->simplifiedTypeRef() ) : nullptr ) )
         {
            x = point->x();
            y = point->y();

            xSuccess = true;
            ySuccess = true;
          }
```

nyalldawson commented on src/ui/qgstextformatwidgetbase.ui:
```diff
@@ -5558,6 +5558,33 @@ font-style: italic;</string>
                             </widget>
                            </item>
+                           <item>
+                            <spacer name="horizontalSpacer_6">
```
Looking at the screenshot this layout has resulted in a very wide horizontal minimum size on the widget. Can you instead move the new Point label and button to a new row below the x/y buttons?

nyalldawson commented on src/core/labeling/qgspallabeling.h:
```diff
@@ -321,6 +321,7 @@ class CORE_EXPORT QgsPalLayerSettings
       // (data defined only)
       PositionX = 9, //!< X-coordinate data defined label position
       PositionY = 10, //!< Y-coordinate data defined label position
+      PositionPoint = 79, //!< Point-coordinate data defined label position
```
```suggestion
      PositionPoint = 114, //!< Point-coordinate data defined label position
```

I vaguely recall that an entry was removed from this enum in 3.0, which would account for the gap. If that's so then we should avoid reusing the possibly conflicting value

nyalldawson commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -244,6 +245,7 @@ void QgsPalLayerSettings::initPropertyDefinitions()
     { QgsPalLayerSettings::PolygonLabelOutside, QgsPropertyDefinition( "PolygonLabelOutside", QgsPropertyDefinition::DataTypeString, QObject::tr( "Label outside polygons" ),  QObject::tr( "string " ) + "[<b>yes</b> (allow placing outside)|<b>no</b> (never place outside)|<b>force</b> (always place outside)]", origin ) },
     { QgsPalLayerSettings::PositionX, QgsPropertyDefinition( "PositionX", QObject::tr( "Position (X)" ), QgsPropertyDefinition::Double, origin ) },
     { QgsPalLayerSettings::PositionY, QgsPropertyDefinition( "PositionY", QObject::tr( "Position (Y)" ), QgsPropertyDefinition::Double, origin ) },
+    { QgsPalLayerSettings::PositionPoint, QgsPropertyDefinition( "PositionPoint", QObject::tr( "Position (point)" ), QgsPropertyDefinition::Point, origin ) },
```
```suggestion
    { QgsPalLayerSettings::PositionPoint, QgsPropertyDefinition( "PositionPoint", QgsPropertyDefinition::DataTypeString, QObject::tr( "Position (point)" ), QObject::tr( "string " ) + "A point geometry", origin ), origin ) },
```

nyalldawson replied:
Because the value **can't** be specified as "[1,2]" -- but I was wrong in suggesting "string" here, it should refer to "geometry". I.e. "Expected format: point geometry value" or similar

nyalldawson commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
+
+    if ( mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionX )
+         && mDataDefinedProperties.isActive( QgsPalLayerSettings::PositionY )
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionX, context.expressionContext() ).isNull()
+         && !mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() ).isNull() )
```
We are double-evaluating these two properties -- can you store the value from the first evaluation to avoid this?

nyalldawson commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
       {
-        exprVal = mDataDefinedProperties.value( QgsPalLayerSettings::PositionY, context.expressionContext() );
-        if ( !exprVal.isNull() )
+        point = QgsPoint( pointAsVariant.value<QgsReferencedGeometry>().asPoint() );
```
We probably **should** be transforming the point here, not ignoring the point's CRS entirely :wink:

domi4484 replied:
I have introduced a warning if the point crs is not equal to the layer crs. Theoretically it is possible to have a different crs for the additional geometry but i can't see practical cases where it is the case.

nyalldawson commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
+      {
+        point = QgsPoint( pointAsVariant.value<QgsGeometry>().asPoint() );
+      }
+      else if ( !pointAsVariant.toString().isEmpty() )
+      {
+        point.fromWkt( pointAsVariant.toString() );
+      }
```
I don't think this logic is correct. I think if a user has a field with WKT, they should use an expression with the geom_from_wkt() function instead (which would mean that the result is always a QgsGeometry)

nyalldawson commented on src/core/labeling/qgspallabeling.cpp:
```diff
@@ -2373,119 +2373,159 @@ std::unique_ptr<QgsLabelFeature> QgsPalLayerSettings::registerFeatureWithDetails
+      {
+        anchorPosition = QgsPointXY( xPos, yPos );
+
+        // only rotate non-pinned OverPoint placements until other placements are supported in pal::Feature
```
this is old --didn't we remove that constraint recently? I suspect there's a merge mistake here which has reintroduced the old code.

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
   return true;
 }
 
+bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess, double &y, bool &ySuccess, int &xCol, int &yCol, int &pointCol ) const
```
Can you extend the existing currentLabelDataDefinedPosition method instead of reimplementing it entirely? This is a LOT of duplicate code

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -678,6 +678,101 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
+  return true;
+}
+
+bool QgsMapToolLabel::layerIsRotatable( QgsVectorLayer *vlayer, int &rotationCol ) const
```
Again, can't we adapt the existing method instead of the duplicate code?

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -859,6 +979,34 @@ bool QgsMapToolLabel::isPinned()
   return rc;
 }
 
+bool QgsMapToolLabel::labelMoveable( QgsVectorLayer *vlayer, const QgsPalLayerSettings &settings, int &xCol, int &yCol, int &pointCol ) const
```
More duplicate code

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -939,7 +1087,7 @@ bool QgsMapToolLabel::createAuxiliaryFields( LabelDetails &details, QgsPalIndexe
       index = vlayer->fields().lookupField( prop.field() );
     }
-    else
+    else if ( prop.propertyType() != QgsProperty::ExpressionBasedProperty )
```
This is handled in QgsAuxiliaryLayer::createProperty -- there's logic in there which upgrades an expression based property to allow for combined handling of expressions AND manually placed labels.

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -990,7 +1138,7 @@ bool QgsMapToolLabel::createAuxiliaryFields( LabelDetails &details, QgsDiagramIn
       index = vlayer->fields().lookupField( prop.field() );
     }
-    else
+    else if ( prop.propertyType() != QgsProperty::ExpressionBasedProperty )
```
```suggestion
    else
```
see above for reason

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -1040,7 +1188,7 @@ bool QgsMapToolLabel::createAuxiliaryFields( QgsCalloutPosition &details, QgsCal
       index = vlayer->fields().lookupField( prop.field() );
     }
-    else
+    else if ( prop.propertyType() != QgsProperty::ExpressionBasedProperty )
```
```suggestion
    else
```
see above for reason

nyalldawson commented on src/app/labeling/qgsmaptoolmovelabel.cpp:
```diff
@@ -221,34 +244,35 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
       }
 
-      if ( xCol >= 0 && yCol >= 0 )
+      bool usesAuxFields = vlayer->fields().fieldOrigin( xCol ) == QgsFields::OriginJoin
+                           && vlayer->fields().fieldOrigin( yCol ) == QgsFields::OriginJoin;
```
Removing the check is unsafe -- what if xCol or yCol is -1 ?

domi4484 committed:
Apply suggestions from code review

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

domi4484 committed:
Apply suggestions from code review

domi4484 committed:
Reverted unneeded changes

domi4484 commented:
I'am finally back to this, thank you very much @nyalldawson for your review last time  🙏

nyalldawson commented:
@domi4484 
> I will make the old function call the new one. I can't extend it because otherwise will break the api (reference parameters can't have default values).

Those classes are in "app", which isn't constrained by the stable api promise. Feel free to break and adapt them as required! :+1:

domi4484 committed:
unified signatures and removed unused function

domi4484 commented:
> Those classes are in "app", which isn't constrained by the stable api promise. Feel free to break and adapt them as required! +1

Good to know, so I cleaned up a little further.
Not sure if it was a good idea but I also removed the function `QgsMapToolLabel::layerCanPin` which wasn't used anywhere.

m-kuhn commented:
@nyalldawson does the "changes requested" still apply here?

m-kuhn commented:
Gentle ping, @nyalldawson

nyalldawson commented:
@m-kuhn I'll look again when I'm back from leave in another week 👍

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -661,18 +662,37 @@ bool QgsMapToolLabel::currentLabelDataDefinedPosition( double &x, bool &xSuccess
+        if ( pointAsVariant.canConvert<QgsGeometry>() )
+        {
+          const  QgsGeometry geometry = pointAsVariant.value<QgsGeometry>();
+          if ( const QgsPoint *point  = ( geometry.constGet() ? qgsgeometry_cast<QgsPoint *>( geometry.constGet()->simplifiedTypeRef() ) : nullptr ) )
```
```suggestion
          if ( const QgsPoint *point  = ( geometry.constGet() ? qgsgeometry_cast<const QgsPoint *>( geometry.constGet()->simplifiedTypeRef() ) : nullptr ) )
```

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -848,17 +859,45 @@ bool QgsMapToolLabel::isPinned()
+  {
+    QString pointColName = dataDefinedColumnName( QgsPalLayerSettings::PositionPoint, settings, vlayer );
+    pointCol = vlayer->fields().lookupField( pointColName );
+    if ( pointCol > 0 )
```
```suggestion
    if ( pointCol >= 0 )
```

nyalldawson commented on src/app/labeling/qgsmaptoollabel.cpp:
```diff
@@ -848,17 +859,45 @@ bool QgsMapToolLabel::isPinned()
+    QString yColName = dataDefinedColumnName( QgsPalLayerSettings::PositionY, settings, vlayer );
+    xCol = vlayer->fields().lookupField( xColName );
+    yCol = vlayer->fields().lookupField( yColName );
+    if ( xCol > 0 || yCol > 0 )
```
```suggestion
    if ( xCol >= 0 || yCol >= 0 )
```

nyalldawson commented on src/app/labeling/qgsmaptoolmovelabel.cpp:
```diff
@@ -182,16 +182,18 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
       {
-        QgsPalIndexes indexes;
+        if ( mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::PositionPoint ) )
+          mCurrentLabel.settings.dataDefinedProperties().property( QgsPalLayerSettings::PositionPoint ).setActive( false );
```
why this change?

domi4484 replied:
When using the auxiliary layer we store the values as X and Y coordinates as it is now

nyalldawson replied:
So is this situation even possible to hit? If so, I don't think we should just silently throw away the user's data defined configuration for PositionPoint. Better to show a message bar warning instead that the label cannot be moved because of .... and abort early here.

domi4484 replied:
Yes it is possible when the point position is defined as a read only expression (not pointing to a writable geometry column).
Ok it make sense to change it like you described.

nyalldawson commented:
Getting very close now!

domi4484 committed:
Apply suggestions from code review

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

domi4484 committed:
Apply suggestions from code review

domi4484 commented on src/app/labeling/qgsmaptoolmovelabel.cpp:
```diff
@@ -221,34 +223,31 @@ void QgsMapToolMoveLabel::cadCanvasPressEvent( QgsMapMouseEvent *e )
         yCol = indexes[ QgsDiagramLayerSettings::PositionY ];
       }
 
-      if ( xCol >= 0 && yCol >= 0 )
```
Check is back again

domi4484 commented:
Thanks again @nyalldawson , I addressed all the remaining points

nyalldawson commented:
Can you fix the merge conflicts and then we can merge...?

domi4484 committed:
Rebase to master

domi4484 committed:
Rebase to master

m-kuhn commented:
:tada:
